File: problems/basic/21/solution1.ts
// Accepted: 53ms (94.88%), 44.85MB (62.25%)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const mergedDummyHead = new ListNode();
    let merged = mergedDummyHead;

    while (list1 !== null && list2 !== null) {
        if (list1.val < list2.val) {
            merged.next = list1;
            merged = merged.next;
            list1 = list1.next;
        } else {
            merged.next = list2;
            merged = merged.next;
            list2 = list2.next;
        }
    }

    while (list1 !== null) {
        merged.next = list1;
        merged = merged.next;
        list1 = list1.next;
    }
    while (list2 !== null) {
        merged.next = list2;
        merged = merged.next;
        list2 = list2.next;
    }

    return mergedDummyHead.next;
};

File: problems/basic/206/solution1.py
# Accepted: 36ms (86.46%), 18.64MB (18.40%)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        cur = head

        while cur is not None:
            oldPrev = prev

            prev = cur
            cur = cur.next

            prev.next = oldPrev

        return prev
